elephant/joint_isi_dithering_class.py

```diff
@@ -0,0 +1,462 @@
+                 dither=15.*pq.ms,
+                 unit=pq.s,
+                 window_length=120.*pq.ms,
+                 num_bins=120,
+                 sigma=1.*pq.ms,
+                 isi_median_threshold=30*pq.ms,
+                 alternate=True,
+                 show_plot=False,
+                 print_mode=False,
+                 use_sqrt=False,
+                 method='fast',
+                 cutoff=True,
+                 min_spikes=10
+                 ):
+        self.st = st
+
+        self.n_surr = n_surr
+
+        self.unit = unit
+        self.dither = dither
+        self.window_length = window_length
+        self.sigma = sigma
+        self.isi_median_threshold = isi_median_threshold
+
+        self.num_bins = num_bins
+
+        self.alternate = alternate
+        self.show_plot = show_plot
+        self.print_mode = print_mode
+        self.use_sqrt = use_sqrt
+        self.method = method
+        self.cutoff = cutoff
+        self.min_spikes = min_spikes
+
+        self.preprocessing()
+
+    def preprocessing(self):
+        '''
+        All preprocessing steps for the joint-ISI dithering are done here.
+
+        So first to checks are done. If they are not passed, self.method is
+        set to 'uniform'. The first one asks for the number of spikes.
+        The second compares the median of the ISI-distribution against a
+        threshold.
+
+        If the method is not 'uniform' the cumulative distribution functions
+        for the Joint-ISI dither process are evaluated.
+
+        If method is 'fast':
+        For each slice of the joint-ISI
+        distribution (parallel to the antidiagonal) a cumulative distr.
+        function is calculated.
+
+        If method is 'window':
+        For each point in the joint-ISI distribution a on the line parallel to
+        the antidiagonal all points up to the dither-parameter are included, to
+        calculate the cumulative distribution function.
+
+        The function has no output, but stores its result inside the class.
+        '''
+        if len(self.st) < self.min_spikes:
+            self.method = 'uniform'
+            return None
+
+        self.isi = stats.isi(self.st.rescale(self.unit).magnitude)
+        isi_median = np.median(self.isi)
+
+        if isi_median > self.isi_median_threshold.rescale(
+                self.unit).magnitude:
+            self.method = 'uniform'
+            return None
+
+        if isinstance(self.dither, pq.Quantity):
+            self.dither = self.dither.rescale(self.unit).magnitude
+        if isinstance(self.window_length, pq.Quantity):
+            self.window_length = self.window_length.rescale(
+                self.unit).magnitude
+        if isinstance(self.sigma, pq.Quantity):
+            self.sigma = self.sigma.rescale(self.unit).magnitude
+
+        self.sampling_rhythm = self.alternate + 1
+
+        self.bin_width = self.window_length/self.num_bins
+
+        def index_to_isi(ind):
```

Comment:
I think in principle you dont need to define such simple function. Maybe just state what you do in a comment?

Reply:
I do it, so that it gets more clear and also that I the 1/2 is added once, and there is no need to remember it again.